topcoffea/modules/dataDrivenEstimation.py

```diff
@@ -0,0 +1,161 @@
+        if self.outHist!=None:  # already some processing has been done, so using what is available
+            self.inhist=self.outHist
+            
+        self.outHist={}
+        if 'SumOfEFTweights' in self.inhist:
+            self.outHist['SumOfEFTweights']=self.inhist['SumOfEFTweights']
+            SumOfEFTweights=self.inhist['SumOfEFTweights']
+            SumOfEFTweights.set_sm()
+            self.smsow = {proc: SumOfEFTweights.integrate('sample', proc).values()[()][0] for proc in SumOfEFTweights.identifiers('sample')}
+
+        for key,histo in self.inhist.items():
+            if key == 'SumOfEFTweights': 
+                # we have already dealt with the sum of weights
+                continue 
+
+            if not len(histo.values()): # histo is empty, so we just integrate over appl and keep an empty histo
+                print(f'[W]: Histogram {key} is empty, returning an empty histo')
+                self.outHist[key]=histo.integrate('appl')
+                continue
+
+            if not self.doDDFakes:
+                # if we are not gonna use the data-driven, then we don't care about the application region, so we get rid of it, and the associated dimension
+                srs=[ histo.integrate('appl',ident) for ident in histo.identifiers('appl') if 'SR' in ident.name]
+                if not len(srs):
+                    raise RuntimeError(f"Histogram {key} does not have any signal region")
+                newhist=srs[0]
+                for h in srs[1:]:
+                    newhist = newhist + h # sum doesnt work for some reason...
+            else:
+
+                # First we are gonna scale all MC processes in the AR by the luminosity
+                # we wont do anything to stuff in the SR because we only want the scaling for the 
+                # prompt subtraction
+                name_regex='(?P<sample>.*)UL(?P<year>.*)'
+                pattern=re.compile(name_regex)
+
+                scale_dict={}
+                for sample in histo.identifiers('sample'):
+                    match = pattern.search(sample.name)
+                    sampleName=match.group('sample')
+                    year=match.group('year')
+                    if not match: 
+                        raise RuntimeError(f"Sample {sample} does not match the naming convention")
+                    if year not in ['16','17','18']:
+                        raise RuntimeError(f"Sample {sample} does not match the naming convention")
+
+                    if self.dataName == sampleName or self.chargeFlipName == sampleName:
+                        continue # We do not scale data or data-driven at all 
+
+                    for appl in histo.integrate('sample',sample).identifiers('appl'):
+                        if 'AR' not in appl.name: # we only care about AR 
+                            continue
+                        smweight = self.smsow[sample.name] if sample.name in self.smsow else 1 # dont reweight central samples
```

Comment:
I'm not sure if I understand this line. Is the purpose to only divide by the sum of weights in the case of an EFT sample using `sample.name in self.smsow` to determine if it's an EFT sample or not? If that's what we're trying to do, I'm wondering if both EFT and central samples will actually be in the `self.smsow` dictionary. The dictionary gets filled inside of this if statement `if 'SumOfEFTweights' in self.inhist`, but I think we make that histogram in all cases (not just in the case of EFT samples). So I'm not sure if this check is doing what it is supposed to do, but I might just be confused about what it's doing.

I think a different way of checking if it's an EFT sample or not (if that's what we're trying to do) might be to just check the number of WCs (using HistEFT's `_nwc`).